pkg/operation/botanist/dns.go

```diff
@@ -17,296 +17,316 @@ package botanist
 }
 
-// DestroyExternalDomainDNSRecord destroys the DNS record for the external cluster domain.
-func (b *Botanist) DestroyExternalDomainDNSRecord(ctx context.Context) error {
-	return b.deleteDNSEntry(ctx, DNSExternalName)
+// DefaultExternalDNSEntry returns DeployWaiter which removes the external DNSEntry.
+func (b *Botanist) DefaultExternalDNSEntry(seedClient client.Client) component.DeployWaiter {
+	return component.OpDestroy(dns.NewDNSEntry(
+		&dns.EntryValues{
+			Name: DNSExternalName,
+		},
+		b.Shoot.SeedNamespace,
+		b.ChartApplierSeed,
+		b.ChartsRootPath,
+		b.Logger,
+		seedClient,
+		nil,
+	))
 }
 
-// DeployAdditionalDNSProviders deploys the additional DNS providers configured in the shoot resource.
-func (b *Botanist) DeployAdditionalDNSProviders(ctx context.Context) error {
-	if b.Shoot.Info.Spec.DNS == nil {
-		return nil
+// DefaultInternalDNSProvider returns the internal DNSProvider if internal DNS is
+// enabled and if not, DeployWaiter which removes the internal DNSProvider.
+func (b *Botanist) DefaultInternalDNSProvider(seedClient client.Client) component.DeployWaiter {
+	if b.NeedsInternalDNS() {
+		return dns.NewDNSProvider(
+			&dns.ProviderValues{
+				Name:       DNSInternalName,
+				Purpose:    DNSInternalName,
+				Provider:   b.Garden.InternalDomain.Provider,
+				SecretData: b.Garden.InternalDomain.SecretData,
+				Domains: &dns.IncludeExclude{
+					Include: []string{b.Shoot.InternalClusterDomain},
+				},
+				Zones: &dns.IncludeExclude{
+					Include: b.Garden.InternalDomain.IncludeZones,
+					Exclude: b.Garden.InternalDomain.ExcludeZones,
+				},
+			},
+			b.Shoot.SeedNamespace,
+			b.ChartApplierSeed,
+			b.ChartsRootPath,
+			b.Logger,
+			seedClient,
+			nil,
+		)
 	}
 
-	var (
-		fns               []flow.TaskFn
-		deployedProviders = sets.NewString()
-	)
-	for i, provider := range b.Shoot.Info.Spec.DNS.Providers {
-		p := provider
-		if p.Primary != nil && *p.Primary {
-			continue
-		}
-		fns = append(fns, func(ctx context.Context) error {
+	return component.OpDestroy(dns.NewDNSProvider(
+		&dns.ProviderValues{
+			Name:    DNSInternalName,
+			Purpose: DNSInternalName,
+		},
+		b.Shoot.SeedNamespace,
+		b.ChartApplierSeed,
+		b.ChartsRootPath,
+		b.Logger,
+		seedClient,
+		nil,
+	))
+}
+
+// DefaultInternalDNSEntry returns DeployWaiter which removes the internal DNSEntry.
+func (b *Botanist) DefaultInternalDNSEntry(seedClient client.Client) component.DeployWaiter {
+	return component.OpDestroy(dns.NewDNSEntry(
+		&dns.EntryValues{
+			Name: DNSInternalName,
+		},
+		b.Shoot.SeedNamespace,
+		b.ChartApplierSeed,
+		b.ChartsRootPath,
+		b.Logger,
+		seedClient,
+		nil,
+	))
+}
+
+// AdditionalDNSProviders returns a map containing a map of DNSProviders with their name as key.
+// Providers which are no longer needed / or in use, DeployWaiter which removes said DNSProvider.
+func (b *Botanist) AdditionalDNSProviders(gardenClient, seedClient client.Client) (map[string]component.DeployWaiter, error) {
```

Comment:
You can a get call with `context.TODO()` inside this func. You can contextify this func.

```suggestion
func (b *Botanist) AdditionalDNSProviders(ctx context.Context, gardenClient, seedClient client.Client) (map[string]component.DeployWaiter, error) {
```